salt/cloud/clouds/scaleway.py

```diff
@@ -0,0 +1,515 @@
+# -*- coding: utf-8 -*-
+'''
+Scaleway Cloud Module
+==========================
+
+.. versionadded:: Beryllium
+
+The Scaleway cloud module is used to interact with your Scaleway BareMetal
+Servers.
+
+Use of this module only requires the ``api_key`` parameter to be set. Set up
+the cloud configuration at ``/etc/salt/cloud.providers`` or
+``/etc/salt/cloud.providers.d/scaleway.conf``:
+
+.. code-block:: yaml
+    scaleway-config:
+      # Scaleway organization and token
+      access_key: 0e604a2c-aea6-4081-acb2-e1d1258ef95c
+      token: be8fd96b-04eb-4d39-b6ba-a9edbcf17f12
+      provider: scaleway
+
+:depends: requests
+'''
+
+from __future__ import absolute_import
+
+import copy
+import json
+import logging
+import pprint
+import time
+
+try:
+    import requests
+except ImportError:
+    requests = None
+
+import salt.config as config
+from salt.exceptions import (
+    SaltCloudNotFound,
+    SaltCloudSystemExit,
+    SaltCloudExecutionFailure,
+    SaltCloudExecutionTimeout
+)
+from salt.ext.six.moves import range
+import salt.utils.cloud
+
+
+log = logging.getLogger(__name__)
+
+
+# Only load in this module if the Scaleway configurations are in place
+def __virtual__():
+    ''' Check for Scaleway configurations.
+    '''
+    if requests is None:
+        return False
+
+    if get_configured_provider() is False:
+        return False
+
+    return True
+
+
+def get_configured_provider():
+    ''' Return the first configured instance.
+    '''
+    return config.is_provider_configured(
+        __opts__,
+        __active_provider_name__ or 'scaleway',
+        ('token',)
+    )
+
+
+def avail_images(call=None):
+    ''' Return a list of the images that are on the provider.
+    '''
+    if call == 'action':
+        raise SaltCloudSystemExit(
+            'The avail_images function must be called with '
+            '-f or --function, or with the --list-images option'
+        )
+
+    items = query(method='images')
+    ret = {}
+    for image in items['images']:
+        ret[image['id']] = {}
+        for item in image:
+            ret[image['id']][item] = str(image[item])
+
+    return ret
+
+
+def list_nodes(call=None):
+    ''' Return a list of the BareMetal servers that are on the provider.
+    '''
+    if call == 'action':
+        raise SaltCloudSystemExit(
+            'The list_nodes function must be called with -f or --function.'
+        )
+
+    items = query(method='servers')
+
+    ret = {}
+    for node in items['servers']:
+        public_ips = []
+        private_ips = []
+        image_id = ''
+
+        if node.get('public_ip'):
+            public_ips = [node['public_ip']['address']]
+
+        if node.get('private_ip'):
```

Comment:
private_ip is always present and there is no plan to change this so just don't branch.